modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerTest.java

```diff
@@ -0,0 +1,405 @@
+                    willCompleteSuccessfully()
+            );
+        }
+
+        awaitActualMetastoreRevision();
+
+        assertStartBuildIndexKeyAbsent(indexId);
+
+        for (int partitionId = 0; partitionId < partitions; partitionId++) {
+            assertPartitionBuildIndexKeyAbsent(indexId, partitionId);
+        }
+
+        assertFalse(indexDescriptor(INDEX_NAME).writeOnly());
+    }
+
+    @Test
+    void testMetastoreKeysAfterDropIndex() throws Exception {
+        createIndex(INDEX_NAME);
+
+        int indexId = indexId(INDEX_NAME);
+
+        dropIndex(INDEX_NAME);
+
+        awaitActualMetastoreRevision();
+
+        assertStartBuildIndexKeyAbsent(indexId);
+
+        for (int partitionId = 0; partitionId < partitions; partitionId++) {
+            assertPartitionBuildIndexKeyAbsent(indexId, partitionId);
+        }
+    }
+
+    @Test
+    void testMetastoreKeysAfterDropIndexForOnlyOnePartition() throws Exception {
+        changePartitionCountInCatalog(1);
+
+        testMetastoreKeysAfterDropIndex();
+    }
+
+    @Test
+    void testMetastoreKeysAfterFinishBuildingOnePartitionAndDropIndex() throws Exception {
+        createIndex(INDEX_NAME);
+
+        int indexId = indexId(INDEX_NAME);
+
+        finishBuildingIndexForPartition(indexId, 0);
+
+        dropIndex(INDEX_NAME);
+
+        awaitActualMetastoreRevision();
+
+        assertStartBuildIndexKeyAbsent(indexId);
+
+        for (int partitionId = 0; partitionId < partitions; partitionId++) {
+            assertPartitionBuildIndexKeyAbsent(indexId, partitionId);
+        }
+    }
+
+    @Test
+    void testMetastoreKeysAfterDropIndexWithLastRemainingPartition() throws Exception {
+        createIndex(INDEX_NAME);
+
+        int indexId = indexId(INDEX_NAME);
+
+        for (int partitionId = 1; partitionId < partitions; partitionId++) {
+            finishBuildingIndexForPartition(indexId, partitionId);
+        }
+
+        dropIndex(INDEX_NAME);
+
+        awaitActualMetastoreRevision();
+
+        assertStartBuildIndexKeyAbsent(indexId);
+
+        for (int partitionId = 0; partitionId < partitions; partitionId++) {
+            assertPartitionBuildIndexKeyAbsent(indexId, partitionId);
+        }
+    }
+
+    @Test
+    void testMetastoreKeysAfterDropIndexWithTwoRemainingPartition() throws Exception {
+        createIndex(INDEX_NAME);
+
+        int indexId = indexId(INDEX_NAME);
+
+        for (int partitionId = 2; partitionId < partitions; partitionId++) {
+            finishBuildingIndexForPartition(indexId, partitionId);
+        }
+
+        dropIndex(INDEX_NAME);
+
+        awaitActualMetastoreRevision();
+
+        assertStartBuildIndexKeyAbsent(indexId);
+
+        for (int partitionId = 0; partitionId < partitions; partitionId++) {
+            assertPartitionBuildIndexKeyAbsent(indexId, partitionId);
+        }
+    }
+
+    private void awaitActualMetastoreRevision() throws Exception {
```

Comment:
Should the name say something about applied revision? As the method seems to wait till applied revision reaches the 'global' revision.

Reply:
What name do you think would be suitable?

Reply:
`awaitTillGlobalMetastoreRevisionIsApplied()`